nixos/modules/tasks/filesystems/overlayfs.nix

```diff
@@ -0,0 +1,147 @@
+{ config, lib, pkgs, utils, ... }:
+
+let
+  # The scripted initrd contains some magic to add the prefix to the
+  # paths just in time, so we don't add it here.
+  sysrootPrefix = fs:
+    if config.boot.initrd.systemd.enable && (utils.fsNeededForBoot fs) then
+      "/sysroot"
+    else
+      "";
+
+  # Returns a service that creates the required directories before the mount is
+  # created.
+  preMountService = _name: fs:
+    let
+      rwName = "rw-${utils.escapeSystemdPath fs.mountPoint}";
+
+      prefix = sysrootPrefix fs;
+
+      upperdir = prefix + fs.overlay.upperdir;
+      workdir = prefix + fs.overlay.workdir;
+    in
+    lib.mkIf (fs.overlay.upperdir != null)
+      {
+        "${rwName}" = {
+          unitConfig = {
+            DefaultDependencies = false;
+            RequiresMountsFor = "${upperdir} ${workdir}";
+          };
+          serviceConfig = {
+            Type = "oneshot";
+            ExecStart = "${pkgs.coreutils}/bin/mkdir -p -m 0755 ${upperdir} ${workdir}";
+          };
+        };
+      };
+
+  overlayOpts = { config, ... }: {
+
+    options.overlay = {
+
+      lowerdir = lib.mkOption {
+        type = lib.types.nullOr (lib.types.listOf lib.types.path);
+        default = null;
+        description = lib.mdDoc ''
+          The list of path(s) to the lowerdir(s).
+
+          To create a writable overlay, you MUST provide and upperdir and a
+          workdir.
+
+          You can create a read-only overlay when you provide multiple (at
+          least 2!) lowerdirs and neither an upperdir nor a workdir.
+        '';
+      };
+
+      upperdir = lib.mkOption {
+        type = lib.types.nullOr lib.types.path;
+        default = null;
+        description = lib.mdDoc ''
+          The path to the upperdir.
+
+          If this is null, a read-only overlay is created using the lowerdir.
+
+          If you set this to some value you MUST also set `workdir`.
+        '';
+      };
+
+      workdir = lib.mkOption {
+        type = lib.types.nullOr lib.types.path;
+        default = null;
+        description = lib.mdDoc ''
+          The path to the workdir.
+
+          This MUST be set if you set `upperdir`.
+        '';
+      };
+
+    };
+
+    config = lib.mkIf (config.overlay.lowerdir != null) {
+      fsType = "overlay";
+      device = lib.mkDefault "overlay";
+
+      options =
+        let
+          rwService = "rw-${utils.escapeSystemdPath config.mountPoint}.service";
+
+          prefix = sysrootPrefix config;
+
+          lowerdir = map (s: prefix + s) config.overlay.lowerdir;
+          upperdir = prefix + config.overlay.upperdir;
+          workdir = prefix + config.overlay.workdir;
+        in
+        [
+          "x-systemd.wanted-by=local-fs.target"
+          "x-systemd.before=local-fs.target"
```

Comment:
Oh, I missed these. These actually shouldn't be here at all. Let the `fstab-generator` take care of making these wanted by the correct target, which isn't `local-fs` in initrd.

Reply:
The issue is, fstab generator doesn't take care of this outside the initrd. If I don't add them, the overlays sometimes don't get mounted at all.

Reply:
No, that's not right; or at least it shouldn't be. Any mount without `noauto` is automatically added to `local-fs.target` by `fstab-generator` in stage 2. And `fstab-generator` in initrd does similarly for `initrd-fs.target`. I'm confident these options shouldn't be there, and if they're not getting mounted, that's a separate issue we need to figure out.

Reply:
I have no idea what the separate issue is or could be. I understand that the generator should add the dependencies but clearly something is not working correctly. However, if I add these dependencies explicitly it works reproducibly.

Reply:
@nikstur Can you show me something where it *doesn't* work consistently without these? I was actually testing something unrelated using this yesterday and it was perfectly fine without these.

Reply:
If you remove these lines and run the test a few times with `--check` you will see that it fails every now and then. It doesn't fail reproducibly every time though. Maybe my kerne (or some other detail of my machine)l is influencing it but then again it *works* reproducibly when I add the explicit `wantedBy`. I don't know why though and it doesn't make any sense to me.